### ISSUE TITLE
bcrypt/node issue

### ISSUE THREAD

**Issue by jbcase** [abernix/meteord#27]
Hey @abernix ,

Thanks for all your work with the Docker image.

I'm currently having some issues when deploying using your image and Meteor Up.

`mup logs` returns the following output:

```
[XXX.XXX.XXX.XXX] npm ERR! Failed at the bcrypt@1.0.3 install script.
[XXX.XXX.XXX.XXX] npm ERR! This is probably not a problem with npm. There is likely additional logging output above.
[XXX.XXX.XXX.XXX]
[XXX.XXX.XXX.XXX] npm ERR! A complete log of this run can be found in:
[XXX.XXX.XXX.XXX] npm ERR!     /root/.npm/_logs/2017-11-09T11_20_55_125Z-debug.log
[XXX.XXX.XXX.XXX] npm WARN meteor-dev-bundle@0.0.0 No description
[XXX.XXX.XXX.XXX] npm WARN meteor-dev-bundle@0.0.0 No repository field.
[XXX.XXX.XXX.XXX] npm WARN meteor-dev-bundle@0.0.0 No license field.
[XXX.XXX.XXX.XXX] npm ERR! code ELIFECYCLE
[XXX.XXX.XXX.XXX] npm ERR! errno 1
[XXX.XXX.XXX.XXX] npm ERR! meteor-dev-bundle@0.0.0 install: `node npm-rebuild.js`
[XXX.XXX.XXX.XXX] npm ERR! Exit status 1
[XXX.XXX.XXX.XXX] npm ERR!
[XXX.XXX.XXX.XXX] npm ERR! Failed at the meteor-dev-bundle@0.0.0 install script.
[XXX.XXX.XXX.XXX] npm ERR! This is probably not a problem with npm. There is likely additional logging output above.
[XXX.XXX.XXX.XXX]
[XXX.XXX.XXX.XXX] npm ERR! A complete log of this run can be found in:
[XXX.XXX.XXX.XXX] npm ERR!     /root/.npm/_logs/2017-11-09T11_20_55_153Z-debug.log
[XXX.XXX.XXX.XXX]
[XXX.XXX.XXX.XXX]> meteor-dev-bundle@0.0.0 install /bundle/bundle/programs/server
[XXX.XXX.XXX.XXX]> node npm-rebuild.js
[XXX.XXX.XXX.XXX]
[XXX.XXX.XXX.XXX]
[XXX.XXX.XXX.XXX]> bcrypt@1.0.3 install /bundle/bundle/programs/server/npm/node_modules/bcrypt
[XXX.XXX.XXX.XXX]> node-pre-gyp install --fallback-to-build
[XXX.XXX.XXX.XXX]
[XXX.XXX.XXX.XXX] sh: 1: node-pre-gyp: not found
[XXX.XXX.XXX.XXX] npm ERR! file sh
[XXX.XXX.XXX.XXX] npm ERR! code ELIFECYCLE
[XXX.XXX.XXX.XXX] npm ERR! errno ENOENT
[XXX.XXX.XXX.XXX] npm ERR! syscall spawn
[XXX.XXX.XXX.XXX] npm ERR! bcrypt@1.0.3 install: `node-pre-gyp install --fallback-to-build`
[XXX.XXX.XXX.XXX] npm ERR! spawn ENOENT
[XXX.XXX.XXX.XXX] npm ERR!
[XXX.XXX.XXX.XXX] npm ERR! Failed at the bcrypt@1.0.3 install script.
[XXX.XXX.XXX.XXX] npm ERR! This is probably not a problem with npm. There is likely additional logging output above.
[XXX.XXX.XXX.XXX]
```

Any idea how do fix this?

Thanks

**Comment by abernix** [comment]
Which image are you using, exactly?

**Comment by abernix** [comment]
Closing as idle.  Happy to discuss further.

**Comment by gjr-qm** [comment]
I'm also having issues with bcrypt trying to install a Meteor app with Meteor up. I used different versions of the image and different versions of Meteor up, but every-time the process ends with:

```
> bcrypt@1.0.3 install /built_app/programs/server/npm/node_modules/bcrypt
> node-pre-gyp install --fallback-to-build

sh: 1: node-pre-gyp: Permission denied
npm ERR! code ELIFECYCLE
npm ERR! errno 126
npm ERR! bcrypt@1.0.3 install: `node-pre-gyp install --fallback-to-build`
npm ERR! Exit status 126
npm ERR!
npm ERR! Failed at the bcrypt@1.0.3 install script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     /root/.npm/_logs/2017-11-23T18_08_42_898Z-debug.log
npm WARN meteor-dev-bundle@0.0.0 No description
npm WARN meteor-dev-bundle@0.0.0 No repository field.
npm WARN meteor-dev-bundle@0.0.0 No license field.
npm ERR! code ELIFECYCLE
npm ERR! errno 126
npm ERR! meteor-dev-bundle@0.0.0 install: `node npm-rebuild.js`
npm ERR! Exit status 126
npm ERR!
npm ERR! Failed at the meteor-dev-bundle@0.0.0 install script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     /root/.npm/_logs/2017-11-23T18_08_42_938Z-debug.log
```
The resulting (intermediate) container keeps restarting...

I tried a lot of things, and I think it has something to do with permissions. What is the best way to find out what causes this problem?

**Comment by gjr-qm** [comment]
One more thing: I don't deploy my app with the root user (read somewhere that deploying with root user using password solved the problem), but with a sudo user (using key to authenticate). Maybe that has something to do with it? I don't want to use root for deployment for obvious reasons...

Als forgot to mention that my machine runs Windows 10.

**Comment by gjr-qm** [comment]
Looks like this error is fixed a few hours ago in Meteor Up (https://github.com/zodern/meteor-up/issues/787) and @zodern will release an update. I will check it out and report my findings here.

Sorry that I thought it was an error in the image (because i tried to manually start it with the bundle prepared by Meteor (up) and that also didn't work).

**Comment by gjr-qm** [comment]
Finally managed to get everything up and running. I had to create my own image that installs node-pre-gyp before the run_app.sh script is executed.

The reason is that we use the accounts package in our Meteor 1.5.1 application. Because of the version we couldn't use the latest meteord image (incompatibilities in node modules). So we used the 4.8.4 binbuild meteord image. But bcrypt didn't install because node-pre-gyp was missing, resulting in the `sh: 1: node-pre-gyp: Permission denied` error. When node-pre-gyp is installed with the install_binbuild_tools everything runs fine.

@abernix maybe add node-pre-gyp to the image (npm i -g node-pregyp) in case it is needed for installation of other node modules?
